services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java

```diff
@@ -0,0 +1,153 @@
+                        }
+                    }
+                    if (recoveryCodeCredentialModel == null || recoveryCodeCredentialModel.allCodesUsed()) {
+                        authenticatedUser.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES);
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    protected boolean isDisabledByBruteForce(AuthenticationFlowContext authnFlowContext, UserModel authenticatedUser) {
+        boolean bolResult = false;
+        String bruteForceError;
+        Response challengeResponse;
+        bruteForceError = AuthenticatorUtils.getDisabledByBruteForceEventError(authnFlowContext.getProtector(),
+                authnFlowContext.getSession(), authnFlowContext.getRealm(), authenticatedUser);
+        if (bruteForceError != null) {
+            authnFlowContext.getEvent().user(authenticatedUser);
+            authnFlowContext.getEvent().error(bruteForceError);
+            challengeResponse = createLoginForm(authnFlowContext, false, Messages.INVALID_USER, FIELD_USERNAME);
+            authnFlowContext.forceChallenge(challengeResponse);
+            bolResult = true;
+        }
+        return bolResult;
+    }
+
+    private Response createLoginForm(AuthenticationFlowContext authnFlowContext, boolean withInvalidUserCredentialsError,
+            String errorToRaise, String fieldError) {
+        Response challengeResponse;
+        LoginFormsProvider loginFormsProvider;
+        if (withInvalidUserCredentialsError) {
+            loginFormsProvider = authnFlowContext.form();
+            authnFlowContext.getEvent().user(authnFlowContext.getUser());
+            authnFlowContext.getEvent().error(Errors.INVALID_USER_CREDENTIALS);
+            loginFormsProvider.addError(new FormMessage(fieldError, errorToRaise));
+        } else {
+            loginFormsProvider = authnFlowContext.form().setExecution(authnFlowContext.getExecution().getId());
+            if (errorToRaise != null) {
+                if (fieldError != null) {
+                    loginFormsProvider.addError(new FormMessage(fieldError, errorToRaise));
+                } else {
+                    loginFormsProvider.setError(errorToRaise);
+                }
+            }
+        }
+        challengeResponse = loginFormsProvider.createLoginRecoveryAuthnCode();
+        return challengeResponse;
+    }
+
+    @Override
+    public boolean requiresUser() {
+        return true;
+    }
+
+    @Override
+    public boolean configuredFor(KeycloakSession session, RealmModel realm, UserModel user) {
+        return session.userCredentialManager().isConfiguredFor(realm, user, RecoveryAuthnCodesCredentialModel.TYPE);
+    }
+
+    @Override
+    public void setRequiredActions(KeycloakSession session, RealmModel realm, UserModel user) {
+        user.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES.name());
```

Comment:
I think it will be better if required action is added to the authentication session rather than to the user. See `OTPFormAuthenticator` or `WebAuthnAuthenticator` classes for the inspiration. Or is there any reason why for backup codes it needs to be done differently than for OTP/WebAuthn?

Reply:
It should be added to auth-session right, otherwise we're updating the user session when that's not needed, and also it would linger in a subsequent login, which isn't correct.

Reply:
What priority should I give the required action for recovery codes? Should it be very low or should it be near where the Configure OTP action is in priority?

Reply:
Hmm.... I am not sure what priority you have in mind? Could you please elaborate? This note was about that instead of calling `user.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES.name())` you can call something like `authenticationSession.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES.name())` to make sure that adding required action is scoped only to this authentication and not to user himself. There is no any priority involved unless I am missing something?

Reply:
I must have commented on the wrong note. I understand what to do about removing user.addRequiredAction.

My question was about the note above that about making the required action default. There is a priority that's a number. Right now, when you register a required action it gives it the lowest priority.

See the MR at https://github.com/rh-iatanaso/keycloak/pull/7/files. In this MR I've set it to 5 but that's almost certainly wrong.